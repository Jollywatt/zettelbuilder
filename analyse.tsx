import { walkSync } from "@std/fs"
import assert from "node:assert"
import * as Path from "@std/path"
import { render } from "preact-render-to-string"

export interface Note {
	name: string
	type: string | null
	dir: Array<string>
	files: { [ext: string]: string }
}

interface NoteTypes {
	[label: string]: Array<string>
}

function detectNoteKind(nt: NoteTypes, extensions: Set<string>): string | null {
	for (const [type, exts] of Object.entries(nt)) {
		if (extensions.symmetricDifference(new Set(exts)).size == 0) {
			return type
		}
	}
	return null
}

export function findNoteFiles(dir: string) {
	const dirnorm = dir.replace(/\/$/, "")
	const iter = walkSync(dirnorm, {
		includeDirs: false,
		match: [/\.note\.\w+$/],
	})

	return Array.from(iter).map((entry) => entry.path)
}

export function notesFromFiles(
	paths: Array<string>,
	{ noteTypes, root = "" }: { noteTypes: NoteTypes; root?: string },
): { [name: string]: Note } {
	const notes: { [name: string]: Note } = {}

	// for each path, extract note name, etc
	for (const path of paths) {
		const parts = Path.parse(path)

		const name = parts.name.replace(/\.note$/, "")
		const rel = Path.relative(root, parts.dir)
		const dir = rel.length ? rel.split(Path.SEPARATOR) : []

		if (!(name in notes)) {
			notes[name] = {
				name,
				dir,
				type: null,
				files: {},
			}
		} else {
			// don't allow notes of same name but different directories
			assert(
				notes[name].dir.join("/") == dir.join("/"),
				`Note name '${name}' used by files in different directories: found ${path} and ${
					Object.values(notes[name].files)
				}.`,
			)
		}

		const ext = parts.ext.slice(1)
		notes[name].files[ext] = path
	}

	// deduce note types from file extensions present
	for (const name in notes) {
		const extensions = Object.keys(notes[name].files)
		const type = detectNoteKind(noteTypes, new Set(extensions))
		if (type === null) {
			console.error(
				`Couldn't determine type of note "${name}" with extensions: ${
					extensions.join(", ")
				}`,
			)
			continue
		}
		notes[name].type = type
	}

	return notes
}

export interface NoteFolder {
	notes: { [name: string]: Note }
	folders: { [name: string]: NoteFolder }
}

export function notesByFolder(notes: { [name: string]: Note }): NoteFolder {
	const tree: NoteFolder = { notes: {}, folders: {} }

	for (const name in notes) {
		const note = notes[name]
		let dir = tree
		for (const c of note.dir) {
			if (dir.folders[c] === undefined) {
				dir.folders[c] = { notes: {}, folders: {} }
				dir = dir.folders[c]
			}
		}
		dir.notes[name] = note
	}

	return tree
}

interface Project {
	noteTypes: NoteTypes
	notes: { [name: string]: Note }
	tree: NoteFolder
	renderPage: Function
}

export function setupProject({
	srcdir,
	sitedir = "build/",
	noteTypes,
}: {
	srcdir: string
	sitedir: string
	noteTypes: NoteTypes
}): Project {
	const files = findNoteFiles(srcdir)
	const notes = notesFromFiles(files, {
		noteTypes: noteTypes,
		root: srcdir,
	})
	const tree = notesByFolder(notes)

	return {
		noteTypes,
		notes,
		tree,
		renderPage: (path: string, page) => {
			const sitepath = Path.join(sitedir, path)
			console.log(`Writing ${sitepath}`)
			Deno.writeTextFileSync(sitepath, render(page))
		},
	}
}
