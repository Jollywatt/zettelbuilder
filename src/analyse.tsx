import { existsSync, walkSync } from "@std/fs"
import assert from "node:assert"
import * as Path from "@std/path"
import { render } from "preact-render-to-string"
import { create as BrowserSync } from "browser-sync"

/**
 * A text file pointer with lazily loaded content.
 *
 * The file's content is read and stored on first request.
 */
class LazyFile {
	path: string
	#content: string | null = null

	constructor(path: string) {
		existsSync(path)
		this.path = path
	}

	get content() {
		if (this.#content === null) {
			console.log(`%cReading%c ${this.path}`, "color: yellow", "")
			this.#content = Deno.readTextFileSync(this.path)
		}
		return this.#content
	}
}

interface NoteInfo {
	name: string
	type: typeof Note
	dir: Array<string>
	files: { [ext: string]: LazyFile }
}

type NoteTypes = Array<typeof Note>

function detectNoteKind(
	nt: NoteTypes,
	extensions: Set<string>,
): typeof Note | undefined {
	for (const noteClass of nt) {
		const extSet = new Set(noteClass.extensionCombo)
		if (extSet.symmetricDifference(new Set(extensions)).size == 0) {
			return noteClass
		}
	}
	return undefined
}

export function findNoteFiles(dir: string) {
	const dirnorm = dir.replace(/\/$/, "")
	const iter = walkSync(dirnorm, {
		includeDirs: false,
		match: [/\.note\.\w+$/],
	})

	return Array.from(iter).map((entry) => entry.path)
}

export function notesFromFiles(
	paths: Array<string>,
	{ noteTypes, srcdir, sitedir = "site/" }: {
		noteTypes: NoteTypes
		srcdir: string
		sitedir?: string
	},
): { [name: string]: Note } {
	const noteInfo: { [name: string]: NoteInfo } = {}

	// for each path, extract note name, etc
	for (const path of paths) {
		const parts = Path.parse(path)

		const name = parts.name.replace(/\.note$/, "")
		const rel = Path.relative(srcdir, parts.dir)
		const dir = rel.length ? rel.split(Path.SEPARATOR) : []

		if (!(name in noteInfo)) {
			noteInfo[name] = {
				name,
				dir,
				type: Note,
				files: {},
			}
		} else {
			// don't allow notes of same name but different directories
			assert(
				noteInfo[name].dir.join("/") == dir.join("/"),
				`NoteInfo name '${name}' used by files in different directories: found ${path} and ${
					Object.values(noteInfo[name].files)
				}.`,
			)
		}

		const ext = parts.ext.slice(1)
		noteInfo[name].files[ext] = new LazyFile(path)
	}

	// deduce note types from file extensions present
	for (const name in noteInfo) {
		const extensions = Object.keys(noteInfo[name].files)
		const type = detectNoteKind(noteTypes, new Set(extensions))
		if (type === undefined) {
			console.error(
				`Couldn't determine type of note "${name}" with extensions: ${
					extensions.join(", ")
				}`,
			)
			continue
		} else {
			noteInfo[name].type = type
		}
	}

	const notes: { [name: string]: Note } = {}
	for (const name in noteInfo) {
		const info = noteInfo[name]
		const NoteClass = info.type
		notes[name] = new NoteClass({
			name: info.name,
			dir: info.dir,
			files: info.files,
			sitedir: sitedir,
		})
	}

	return notes
}

export interface NoteFolder {
	notes: { [name: string]: Note }
	folders: { [name: string]: NoteFolder }
}

export function notesByFolder(
	notes: { [name: string]: Note },
): NoteFolder {
	const tree: NoteFolder = { notes: {}, folders: {} }

	for (const name in notes) {
		const note = notes[name]
		let dir = tree
		for (const c of note.dir) {
			if (dir.folders[c] === undefined) {
				dir.folders[c] = { notes: {}, folders: {} }
			}
			dir = dir.folders[c]
		}
		dir.notes[name] = note
	}

	return tree
}

export class Note {
	static extensionCombo: string[] = []

	name: string
	dir: string[]
	files: { [extension: string]: LazyFile }
	sitedir: string

	static description: string | null = null
	get description() {
		const staticDescription = (this.constructor as typeof Note).description
		if (staticDescription !== null) return staticDescription
		return Object.keys(this.files).join(", ")
	}

	constructor({
		name,
		dir,
		files,
		sitedir,
	}) {
		this.name = name
		this.dir = dir
		this.files = files
		this.sitedir = sitedir
	}

	render() {
		return (
			<main>
				This page was generated by the default note renderer.
				<pre>{Deno.inspect(this)}</pre>
			</main>
		)
	}
}

export interface ProjectData {
	files: Array<string>
	notes: { [name: string]: Note }
	tree: NoteFolder
}

export class Project {
	srcdir: string
	sitedir: string
	noteTypes: NoteTypes
	builder: Function

	analysis: ProjectData | null

	constructor({
		srcdir,
		sitedir,
		noteTypes,
		builder,
	}: {
		srcdir: string
		sitedir: string
		noteTypes: NoteTypes
		builder: Function
	}) {
		this.srcdir = srcdir
		this.sitedir = sitedir
		this.noteTypes = noteTypes
		this.builder = builder
		this.analysis = null
	}

	analyse() {
		const files = findNoteFiles(this.srcdir)
		const notes = notesFromFiles(files, {
			noteTypes: this.noteTypes,
			srcdir: this.srcdir,
			sitedir: this.sitedir,
		})
		const tree = notesByFolder(notes)

		this.analysis = {
			files,
			notes,
			tree,
		}

		return this.analysis
	}

	renderPage(path: string, page) {
		const sitepath = Path.join(this.sitedir, path)
		console.log(`%cWriting%c ${sitepath}`, "font-weight: bold", "")
		const html = typeof page === "string"
			? page
			: `<!DOCTYPE html>` + render(page, { pretty: true })
		Deno.writeTextFileSync(sitepath, html)
	}

	build() {
		console.log(`%cBuilding%c site at ${this.sitedir}`, "font-weight: bold", "")
		// ensure site directory exists and is empty
		Deno.mkdirSync(this.sitedir, { recursive: true })
		Deno.removeSync(this.sitedir, { recursive: true })
		Deno.mkdirSync(this.sitedir)
		this.builder(this)
	}

	serve(
		{ autoreload = true, autobuild = true } = {
			autoreload: Boolean,
			autobuild: Boolean,
		},
	) {
		const bs = BrowserSync()

		if (autoreload) {
			bs.watch(Path.join(this.sitedir, "**/*.{html,css}")).on(
				"change",
				bs.reload,
			)
		}

		bs.watch(Path.join(this.srcdir, "**/*")).on("change", (path) => {
			console.log(`Detected change: ${path}`)
			this.build()
		})

		bs.watch(Deno.mainModule, (path) => {
			console.log(`Detected change in zettelsite package: ${path}`)
			this.build()
		})

		bs.init({
			server: this.sitedir,
			serveStatic: [this.sitedir],
			serveStaticOptions: {
				extensions: ["html"], // pretty urls
			},
			open: false,
		})
	}
}
