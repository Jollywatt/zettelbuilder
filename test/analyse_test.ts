import { assertEquals } from "@std/assert"
import { assertSnapshot } from "@std/testing/snapshot"

import { findNoteFiles, notesFromFiles } from "../src/analyse.tsx"

Deno.test("note detection", async (shot) => {
	await assertSnapshot(
		shot,
		notesFromFiles([
			`maths/a.note.md`,
			`maths/b.note.md`,
			`c.note.typ`,
			`c.note.pdf`,
		], {
			noteTypes: {
				"Markdown": ["md"],
				"Typst PDF": ["typ", "pdf"],
			},
		}),
	)
})
