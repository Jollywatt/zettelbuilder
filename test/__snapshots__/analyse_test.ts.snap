export const snapshot = {};

snapshot[`note detection 1`] = `
{
  a: {
    dir: [
      "maths",
    ],
    files: {
      md: "maths/a.note.md",
    },
    name: "a",
    type: "Markdown",
  },
  b: {
    dir: [
      "maths",
    ],
    files: {
      md: "maths/b.note.md",
    },
    name: "b",
    type: "Markdown",
  },
  c: {
    dir: [],
    files: {
      pdf: "c.note.pdf",
      typ: "c.note.typ",
    },
    name: "c",
    type: "Typst PDF",
  },
}
`;
